# This notebook uses the UCI banknote dataset
- https://archive.ics.uci.edu/ml/datasets/banknote+authentication

In [8]:
from sklearn import tree
import graphviz
import os
import numpy as np
from sklearn.model_selection import GridSearchCV
import pandas as pd
from IPython.display import IFrame

## Data set

In [9]:
data = pd.read_csv('../dataset/data_banknote_authentication.csv')
feats = list(data)[:-1]
data.drop_duplicates(keep='first', inplace=True, ignore_index=True, subset=list(data)[:-1])
neg, pos = [data.loc[data['class'] == arg] for arg in (0, 1)]
total = len(data)
print("negative: {:.0%}, positive: {:.0%}".format(len(neg) / total, len(pos) / total))
data.describe()

negative: 55%, positive: 45%


variance     skewness     curtosis      entropy        class
count  1348.000000  1348.000000  1348.000000  1348.000000  1348.000000
mean      0.445785     1.909039     1.413578    -1.168712     0.452522
std       2.862906     5.868600     4.328365     2.085877     0.497925
min      -7.042100   -13.773100    -5.286100    -8.548200     0.000000
25%      -1.786650    -1.627000    -1.545600    -2.393100     0.000000
50%       0.518735     2.334150     0.605495    -0.578890     0.000000
75%       2.853250     6.796025     3.199800     0.403863     1.000000
max       6.824800    12.951600    17.927400     2.449500     1.000000

## Decision Tree

### Gini impurity measure
- Let $p_{nc}$ denote the density of a class $c$ in a given node $n$
$$G(N_n) = \sum\limits_{c} p_{nc}(1-p_{nc})$$

- Minimizing $G$ encourages the model to make splits such that the data is better separated. When $p_{nc}$ approaches 0, then less instances of $c_i$ exist in $n$ and more instances of $c_j$ exist in $n$. The converse is true as well, that is, when $p_{nc}$ approaches 1, then more instances of $c_j$ exist in $n$ and less instances of $c_i$ exist in $n$. When $p_{nc}$ approaches 0 or 1, then the summation term approaches 0 and the summation approaches 0. Thus, the model is encouraged to choose splits that further segregate the data.

### Cost complexity pruning for binary tree
- Cost complexity pruning is used to regularize the tree and prevent overfitting. Cost complexity pruning can be defined as
$$V = C + \alpha * S$$
where $S$ is a sequence $N, N - 2, N - 4, N - 6, ... N - 2 * N_{int}$ and $N_{int}$ is the number of internal nodes in $N$, parameter $\alpha \geq 0$, and $C$ is a cumulative sum of the sequence $0, C_1, C_1, C_2, ... C_n$. $argmin (V)$ gives the appropriate number of prunes to perform. This can be seen by the fact that as $\alpha$ grows, so does $S$. Therefore, more cost must be accrued to reduce the tree complexity and $argmin(V)$ grows accordingly.
- The prune complexity of a single node can be defined as
$$ P_c = W(t) - W(I_t) - W(J_t) $$
where $t$ is some tree, $I$ and $J$ are children of $t$, and $W$ is the sample weighted impurity of the node. $C$ is a sequence of $P_c$ ordered from least costly to most, and the least costly nodes are pruned first until $argmin(V)$ is surpassed.

In [10]:
#Create mapping and model
xs, ys = np.asarray(data[list(data)[:-1]]), np.asarray(data[list(data)[-1]])
model = tree.DecisionTreeClassifier()

#CV multiple models, grid search over cost complexity
param_space = {'ccp_alpha': np.linspace(0.001, 0.05, num=25)}
search = GridSearchCV(estimator=model, param_grid=param_space, cv=5)
search.fit(xs, ys)
df = pd.DataFrame(search.cv_results_)[['param_ccp_alpha', 'rank_test_score', 'std_test_score', 'mean_test_score', 'params']]\
    .sort_values(by=['rank_test_score'])
#print cross validation results sorted by rank_test_score
print(df[list(df)[:-1]])

   param_ccp_alpha  rank_test_score  std_test_score  mean_test_score
0            0.001                1        0.006191         0.985169
1       0.00304167                2        0.007764         0.981457
2       0.00508333                3        0.011463         0.977756
3         0.007125                4        0.011602         0.975534
4       0.00916667                5        0.016964         0.965163
5        0.0112083                6        0.015249         0.957739
6          0.01325                7        0.014021         0.955509
7        0.0152917                8        0.008930         0.946595
8        0.0173333                9        0.018816         0.938439
9         0.019375               10        0.017344         0.929533
11       0.0234583               11        0.015947         0.927311
12          0.0255               11        0.015947         0.927311
10       0.0214167               11        0.015947         0.927311
13       0.0275417               1

- ccp_alpha of 0.05 was used to see how the tree changes. Obviously, it is too aggresive and causes severe underfitting.
- No visiblity on the trees used in CV. However, we can fit the trees again to the entire dataset during analysis and see how pruning affects the shape.
- Accuracy measurements do not correspond one-to-one between succeeding trees and values in table, but they would be close.

In [11]:
#This function will be used to draw decision trees
def graph(model, feats, plot_dir, name):
    dot_data = tree.export_graphviz(model, out_file=None, feature_names=feats, 
                                    class_names=['authentic', 'fraudulent'], 
                                    filled=True, rounded=True, special_characters=True, 
                                    leaves_parallel=True)
    graph = graphviz.Source(dot_data)
    out_path = os.path.join(plot_dir, name)
    graph.render(out_path, format='png', cleanup=True)
    return out_path

In [12]:
overfit_model = tree.DecisionTreeClassifier(**df['params'].iloc[0])
overfit_model.fit(xs, ys)
graph(overfit_model, feats, './graphs', 'overfit')
#Reminder: you can click on the iframe to zoom in
IFrame("./graphs/overfit.png", width=800, height=300)

- This tree overfits the data. There are several buckets that have one sample in them and the tree is too complex. Thus, this tree will not generalize well to unseen data.

In [13]:
underfit_model = tree.DecisionTreeClassifier(**df['params'].iloc[-1])
underfit_model.fit(xs, ys)
graph(underfit_model, feats, './graphs', 'underfit')
IFrame("./graphs/underfit.png", width=800, height=400)

- Choosing the last hyperparameter shows a tree that underfits because it is lacking in complexity. However, I am amazed to see that we can achieve roughly 88% - 91% with such a simple tree.
- If such a simple tree achieves such high accuracy, then that must speak to the high degree of separability in the data set.

In [14]:
bestfit_model = tree.DecisionTreeClassifier(**df['params'].iloc[3])
bestfit_model.fit(xs, ys)
graph(bestfit_model, feats, './graphs', 'bestfit_model')
IFrame("./graphs/bestfit_model.png", width=800, height=400)

- The tree at index 3 appears to be far less complex than the one at 0 and has no small buckets.
- This tree has a higher variance than the tree at index 0, and looks to produce roughly 96% - 98% accuracy.